plugins/module_utils/default_gateway.py

```diff
@@ -0,0 +1,185 @@
+            my_defaultgw = self._gw2machine(params['name'])
+            if params['ipprotocol'] == "inet":
+                obj['defaultgw4'] = my_defaultgw
+            elif params['ipprotocol'] == "inet6":
+                obj['defaultgw6'] = my_defaultgw
+            else:
+                self.module.fail_json(msg='Please specify a valid ipprotocol (inet6/inet4)')
+
+        return obj
+
+    def _validate_params(self):
+        """ do some extra checks on input parameters
+        name        required, str
+        ipprotocol  default : inet, choice inet/inet6
+        """
+        params = self.params
+        gateway_list = ["none", "automatic"] + [gw["Name"] for gw in self.pfsense.find_active_gateways()]
+
+        # get list of current default gateways and append gateway_groups to list
+        for elt in self.root_elt:
+            if elt.tag in ["gateway_group"]:
+                gateway_list.append(elt.find("name").text)
+            elif elt.tag == "defaultgw4":
+                self.defaultgw4 = elt.text
+            elif elt.tag == "defaultgw6":
+                self.defaultgw6 = elt.text
+
+        if params["name"]:
+            if str(params["name"]) not in gateway_list:
+                self.module.fail_json(msg="Unknown gateway %s : %s" % (params["name"], gateway_list))
+
+    ##############################
+    # XML processing
+    #
+    def _create_target(self):
+        """ create the XML target_elt """
+        if self.params["ipprotocol"] == "inet":
+            return self.pfsense.new_element('defaultgw4')
+        elif self.params["ipprotocol"] == "inet6":
+            return self.pfsense.new_element('defaultgw6')
+
+    ##############################
+    # Utilities
+    #
+
+    @staticmethod
+    def _gw2machine(gateway):
+        """
+        Translates special gateway to machine-readable
+        "-" means None
+        "" means Automatic
+        """
+        if gateway is not None:
+            if gateway.lower() == "automatic":
+                return ""
+            elif gateway.lower() == "none":
+                return "-"
+        return gateway
+
+    @staticmethod
+    def _gw2human(gateway):
+        """
+        Translates special gateway as human-readable
+        "-" means None
+        "" means Automatic
+        """
+        if gateway is None:
+            return "automatic"
+        elif gateway == "-":
+            return "none"
+        else:
+            return gateway
+
+    @staticmethod
+    def _get_params_to_remove():
+        """ returns the list of params to remove if they are not set """
+        return []
+
+    ##############################
+    def run(self, params):
+        """ process input params to add/update/delete """
+        self.params = params
+        self._check_deprecated_params()
+        self._check_onward_params()
+        self._validate_params()
+
+        self.obj = self._params_to_obj()
+
+        if params["name"]:
+            self._add()
+        else:
```

Comment:
Let's always return these items.

Reply:
It's done.. I removed the self.defaultgw6/gw4 properties and set the return value elsewhere.. 
I tested it. I worked for me..